### PR TITLE
Add .NET Core 3.1 targeting and package generation

### PR DESCRIPTION
**Pull request by erikma**
NetCore3.1 in VS 16.4 adds C++/CLI support for .NET Core. Add a second 3.1 targeting to the Framework version:

- Add a NetCore 3.1 target for ProjFS.Managed by factoring out common vcxproj portions and creating two distinct vcxproj files with different targets.
- Dual-build the test and SimpleProviderManaged projects.
- New package version translated to SemVer at 1.2.0.0 to supersede existing 1.1.* versions.
- Update the master build script to properly import the VS 16.4+ environment to make MC++ compilation work.
- Update the test script to run the test suite under both target frameworks.

Issue #38

### CODE REVIEW

File: ProjectedFSLib.Managed.props [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <Project ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
 
   <PropertyGroup Label="Parameters">
-    <ProjFSManagedVersion>0.2.173.2</ProjFSManagedVersion>
+    <ProjFSManagedVersion>0.3.0.0</ProjFSManagedVersion>
```
Comment (wilbaker) [inline ProjectedFSLib.Managed.props]: What was the reason for this version number change?  

`0.2.173.2` has been used as a special "built on dev box" version and so unless there is a reason it has to change could we leave it `0.2.173.2`?

File: ProjectedFSLib.Managed.sln [hunk 2]
```diff
@@ -22,24 +24,52 @@ Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Scripts", "Scripts", "{80F7
 		scripts\RunTests.bat = scripts\RunTests.bat
 	EndProjectSection
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ProjectedFSLib.Package", "NugetPackage\ProjectedFSLib.Package.csproj", "{01FCBE2D-7018-42C1-B617-EF02032A889C}"
+EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug|Any CPU = Debug|Any CPU
```
Comment (wilbaker) [inline ProjectedFSLib.Managed.sln]: Was adding `Debug|Any` and `Release|Any` required as part of the move to .NET Core?  @cgallred is ProjFS supported on non-x64 platforms?

File: ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props [hunk 3]
```diff
@@ -1,157 +1,148 @@
...
+    <ProjectConfiguration Include="Release|x64">
+      <Configuration>Release</Configuration>
+      <Platform>x64</Platform>
+    </ProjectConfiguration>
+  </ItemGroup>
+  <PropertyGroup Label="Globals">
+    <VCProjectVersion>16.0</VCProjectVersion>
+    <Keyword>ManagedCProj</Keyword>
+    <RootNamespace>Microsoft.Windows.ProjFS</RootNamespace>
+    <WindowsTargetPlatformVersion>10.0</WindowsTargetPlatformVersion>
```
Comment (wilbaker) [inline ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props]: Does this mean the latest Windows 10 will be targeted?  Could this cause problems if the managed API is deployed on a older version of Windows 10?

File: ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props [hunk 4]
```diff
@@ -1,157 +1,148 @@
...
+    <RunCodeAnalysis>true</RunCodeAnalysis>
+    <TargetName>ProjectedFSLib.Managed</TargetName>
+    <IncludePath>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362\um;$(IncludePath)</IncludePath>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <LinkIncremental>false</LinkIncremental>
+    <CodeAnalysisRuleSet>MixedRecommendedRules.ruleset</CodeAnalysisRuleSet>
+    <RunCodeAnalysis>false</RunCodeAnalysis>
+    <TargetName>ProjectedFSLib.Managed</TargetName>
+    <IncludePath>C:\Program Files (x86)\Windows Kits\10\Include\10.0.18362.0\um;$(IncludePath)</IncludePath>
```
Comment (wilbaker) [inline ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props]: Similar question here, by using `10.0.18362.0` is there any risk that we'll use APIs not supported in `10.0.17763.0`?

File: ProjectedFSLib.Managed.Test/Helpers.cs [hunk 5]
```diff
@@ -13,11 +13,9 @@ namespace ProjectedFSLib.Managed.Test
 {
     class Helpers
     {
-        private Process m_providerProcess;
-        public Process ProviderProcess { get => m_providerProcess; set => m_providerProcess = value; }
+        public Process ProviderProcess { get; set; }
```
Comment (wilbaker) [inline ProjectedFSLib.Managed.Test/Helpers.cs]: Minor suggestion:  These code cleanup changes appear to be unrelated to the move to .NET Core 3.1.

Could they be put in a separate PR to make it easier to see what changes were specifically required for the move to .NET Core?

File: simpleProviderManaged/Program.cs [hunk 6]
```diff
@@ -1,29 +1,47 @@
...
+                    .WithParsed((ProviderOptions options) => Run(options));
 
-            Log.Information("Exit");
+                Log.Information("Exit successfully");
+                return (int) ReturnCode.Success;
+            }
+            catch (Exception ex)
+            {
+                Console.Error.WriteLine($"Unexpected exception: {ex}");
+                Thread.Sleep(5 * 1000);
```
Comment (wilbaker) [inline simpleProviderManaged/Program.cs]: Why is there a call to `Thread.Sleep`?

**Comment by abhijeeg** [comment]
Should we target both .Net Core and .Net Framework as seems to be the case in this PR ? @wilbaker, @cgallred - What's your take on this ?

> Upgrade the NetFramework version to 4.7.2 minimum.
> Add a NetCore 3.1 target for ProjFS.Managed by factoring out common vcxproj portions and creating two distinct vcxproj files with different targets.

**Comment by wilbaker** [comment]
> Should we target both .Net Core and .Net Framework as seems to be the case in this PR ? @wilbaker, @cgallred - What's your take on this ?

If we target .Net Core and .Net Framework then customers will not be forced to move to .NET Core (though they may need to move to a later version of .NET Framework).

Unless there is a large cost to supporting both I don't see a reason to force people onto .NET Core.  @mjcheetham, what are your thoughts?

File: ProjectedFSLib.Managed.sln (hunk 2 above)
Comment (cgallred) [reply ProjectedFSLib.Managed.sln]: @wilbaker ProjFS is available on x64 and x86 only.

File: ProjectedFSLib.Managed.Test/Helpers.cs (hunk 5 above)
Comment (cgallred) [reply ProjectedFSLib.Managed.Test/Helpers.cs]: Agreed.  Please remove all changes that are not strictly required to move to .NET core 3.1.

**Comment by mjcheetham** [comment]
> If we target .NET Core and .NET Framework then customers will not be forced to move to .NET Core (though they may need to move to a later version of .NET Framework).
> 
> Unless there is a large cost to supporting both I don't see a reason to force people onto .NET Core. @mjcheetham, what are your thoughts?

Agreed. If it's not too much effort, keeping a .NET Framework target would be helpful for consumers still targeting Framework.

File: ProjectedFSLib.Managed.props (hunk 1 above)
Comment (erikma) [reply ProjectedFSLib.Managed.props]: Reverted and added a comment, it's specialness was not obvious.

File: simpleProviderManaged/Program.cs (hunk 6 above)
Comment (erikma) [reply simpleProviderManaged/Program.cs]: Accidentally left behind from debugging. Reverted.

File: ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props (hunk 3 above)
Comment (erikma) [reply ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props]: No, it's more related to the installed SDK included with VS.

File: ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props (hunk 4 above)
Comment (erikma) [reply ProjectedFSLib.Managed.API/ProjectedFSLib.Managed.props]: No, this is simply an update to map to the default WinSDK included in VS2019 16.4 for ease of development. The APIs in use here are from RS1 and are not modified by use of a later SDK.

**Comment by erikma** [comment]
RE: NetFramework vs. NetCore: My strong opinion is we should target both. As it stands most projects I'm working on will take years to move to NetCore simply because of the effort needed to move dependencies (like this one!) to Core or Standard one by one then update package references and retest end-to-end. Hundreds of packages at differing Framework vs. Core vs. Standard versions mean a lot of pain. Framework support should never go away in this pakcage until Framework itself is removed from Windows (meaning: effectively never).

**Comment by cgallred** [comment]
@erikma I assume you changed the build script to require >= VS 16.4 because that has .NET Core 3.1, correct?  That is what is breaking the validation checks.  The Azure Pipelines pool ("Hosted Windows 2019 with VS2019") has "VisualStudio/16.3.6+29418.71".  The build script is failing because it can't find a VS 16.4 or greater.

**Comment by erikma** [comment]
@cgallred Correct on the 16 4 requirement, it is a strong requirement. I bet the 2019 image gets the update very soon, as I was able to successfully update to netcore31 in another project in mseng and everything was fine. IIRC they usually test there (ring 0) then roll outward.

**Comment by cgallred** [comment]
> I bet the 2019 image gets the update very soon
@erikma okay, we'll re-submit the build/test periodically.  Once it passes we'll complete the PR.

**Comment by cgallred** [comment]
@wilbaker the 2019 Azure image has been updated to VS 16.4 and the tests are now passing. I've approved the PR but I'd appreciate somebody with more .NET experience putting their stamp of approval on it before I merge. :-)

File: NugetPackage/Microsoft.Windows.ProjFS.nuspec [hunk 7]
```diff
@@ -0,0 +1,20 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
+  <metadata>
+    <id>Microsoft.Windows.ProjFS</id>
+    <version>1.2.0.0</version>
```
Comment (wilbaker) [inline NugetPackage/Microsoft.Windows.ProjFS.nuspec]: Will something automatically update this version as there are new builds?  Or will it need to be updated by hand?

File: NugetPackage/ProjectedFSLib.Package.csproj [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+        OverwriteReadOnlyFiles="true" />
+    <Copy
+        SourceFiles="@(DocFilesToCopy)"
+        DestinationFolder="$(PkgDir)lib\netcoreapp3.1"
+        SkipUnchangedFiles="true"
+        OverwriteReadOnlyFiles="true" />
+
+    <Copy
+        SourceFiles="@(FrameworkFilesToCopy)"
+        DestinationFolder="$(PkgDir)lib\net472"
```
Comment (wilbaker) [inline NugetPackage/ProjectedFSLib.Package.csproj]: I would *guess* VFS4G won't take a new drop of the ProjFS Managed API until it is converted to .NET Core, but if it needed to take a new drop before then would it be able to?

VFS4G targets .NET Framework 4.6.2, would net472 be compatible?

cc: @jrbriggs @mjcheetham

File: ProjectedFSLib.Managed.sln [hunk 9]
```diff
@@ -32,14 +36,22 @@ Global
...
+		{B4018C7F-BF11-47BC-8770-72B05C9437EE}.Release|x64.ActiveCfg = Release|Any CPU
+		{B4018C7F-BF11-47BC-8770-72B05C9437EE}.Release|x64.Build.0 = Release|Any CPU
+		{5697F978-E1ED-4C2E-8218-4110F5EA559D}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{5697F978-E1ED-4C2E-8218-4110F5EA559D}.Debug|x64.Build.0 = Debug|Any CPU
+		{5697F978-E1ED-4C2E-8218-4110F5EA559D}.Release|x64.ActiveCfg = Release|Any CPU
+		{5697F978-E1ED-4C2E-8218-4110F5EA559D}.Release|x64.Build.0 = Release|Any CPU
+		{01FCBE2D-7018-42C1-B617-EF02032A889C}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{01FCBE2D-7018-42C1-B617-EF02032A889C}.Debug|x64.Build.0 = Debug|Any CPU
+		{01FCBE2D-7018-42C1-B617-EF02032A889C}.Release|x64.ActiveCfg = Release|Any CPU
+		{01FCBE2D-7018-42C1-B617-EF02032A889C}.Release|x64.Build.0 = Release|Any CPU
```
Comment (wilbaker) [inline ProjectedFSLib.Managed.sln]: > ProjFS is available on x64 and x86 only.

Where did we land on the "Any CPU" option? Can/should that be removed?

File: scripts/RunTests.bat [hunk 10]
```diff
@@ -3,8 +3,19 @@ CALL %~dp0\InitializeEnvironment.bat || EXIT /b 10
 
 IF "%1"=="" (SET "Configuration=Debug") ELSE (SET "Configuration=%1")
 
-set RESULT=0
+set RESULT_FRAMEWORK=0
+set TESTDIR=%PROJFS_OUTPUTDIR%\ProjectedFSLib.Managed.Test\bin\AnyCPU\%Configuration%\net472
```
Comment (wilbaker) [inline scripts/RunTests.bat]: .NET Question: Do `AnyCPU` binaries work on any platform that supports `net472`?

File: simpleProviderManaged/Program.cs [hunk 11]
```diff
@@ -9,21 +9,37 @@ namespace SimpleProviderManaged
 {
     public class Program
     {
-        public static void Main(string[] args)
+        private enum ReturnCode
```
Comment (wilbaker) [inline simpleProviderManaged/Program.cs]: Nit: It doesn't look like the changes in this file were required to move to .NET Core.  Move them out of this PR?

**Comment by erikma** [comment]
@wilbaker RE Any CPU: See comment in resolved conversation, we have to leave it since VS locks up on trying to remove. The packaging target needs AnyCPU, and the test and main assembly targets are all explicitly set to x64 in each proj so the sln targeting doesn't matter anyway

**Comment by wilbaker** [comment]
>@wilbaker RE Any CPU: See comment in resolved conversation, we have to leave it since VS locks up on trying to remove. The packaging target needs AnyCPU, and the test and main assembly targets are all explicitly set to x64 in each proj so the sln targeting doesn't matter anyway

@erikma thanks!  I am having trouble finding the conversation you mentioned, but the details you included in this comment sound good to me.

File: NugetPackage/Microsoft.Windows.ProjFS.nuspec (hunk 7 above)
Comment (cgallred) [reply NugetPackage/Microsoft.Windows.ProjFS.nuspec]: I hadn't noticed the addition of this file.  There's a [separate pipeline](https://dev.azure.com/microsoft/OS/_build?definitionId=37476&_a=summary) that builds, signs, and packages the manged API.  It automatically sets the version.  This pipeline is triggered when we merge into a branch called `release/*`.

Instead of adding a static nuspec here that pipeline needs to be updated.  Or it needs to be updated to be able to dynamically update the version in this nuspec (if that's possible).

Comment (cgallred) [reply NugetPackage/Microsoft.Windows.ProjFS.nuspec]: I've had a look through the signing pipeline to refresh my memory.  Packaging like this won't work with the workflow that we have.  The pipeline that is triggered by PRs for merging to master doesn't have signing support by design.  It's only when we're merging to a release branch that we want to run the signing pipeline.

That pipeline needs to be updated to find and sign both the Framework and Core outputs.  Unless there's a reason to have this packaging project other than for distribution, this project should be removed.

Comment (cgallred) [reply NugetPackage/Microsoft.Windows.ProjFS.nuspec]: So... I updated the signing pipeline to pick up and sign the Framework and Core outputs and package them up together but hit a snag.  The Package ES pool that the pipeline runs on doesn't have VS 16.4.  It is still running 15.9.  I'll need to talk to the Package ES folks to see if they have a pool w/ the VS version we need.

Comment (cgallred) [reply NugetPackage/Microsoft.Windows.ProjFS.nuspec]: @erikma the signing pipeline has been updated and now passes. @wilbaker that takes care of this version number issue.

File: NugetPackage/ProjectedFSLib.Package.csproj (hunk 8 above)
Comment (cgallred) [reply NugetPackage/ProjectedFSLib.Package.csproj]: @wilbaker @jrbriggs @mjcheetham 
I'm ready to approve this PR now that our validation and build/sign/package pipelines are updated to work with these changes.  Are you okay with these changes?  For instance, is the update from Framework 4.6.2 to 4.7.2 okay?

Comment (jrbriggs) [reply NugetPackage/ProjectedFSLib.Package.csproj]: @cgallred Any reason we can't keep targeting Framework 4.6.2 so we can avoid having to do extra setup work on RS1?

Comment (cgallred) [reply NugetPackage/ProjectedFSLib.Package.csproj]: I don't know if there was a reason to upgrade the Framework version.  @erikma?

**Comment by erikma** [comment]
@wilbaker Latest commit downgrades to net461 to match previous

**Comment by erikma** [comment]
I hadn’t known you still had an RS1 deployment targeting, net472 is what I’ve moved all our other projects to until we can get to netcore. I pulled it back to net461 as it was before.

**Comment by cgallred** [comment]
@wilbaker @jrbriggs @mjcheetham
Erik reverted to Framework 4.6.1.  Do these changes look good to you now?

**Comment by wilbaker** [comment]
>@wilbaker @jrbriggs @mjcheetham
Erik reverted to Framework 4.6.1. Do these changes look good to you now?

These look good to me

**Comment by cgallred** [comment]
@wilbaker @jrbriggs @mjcheetham 

FYI this has been [uploaded to NuGet.org](https://www.nuget.org/packages/Microsoft.Windows.ProjFS/1.2.19351.1).  Thanks again to @erikma for doing this work.

Cc: @abhijeeg @PingXie
